File: src/main/java/one/pkg/modpublish/version/constraint/VersionConstraintParser.java
package one.pkg.modpublish.version.constraint;

import one.pkg.modpublish.version.Version;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class VersionConstraintParser {
    private static final String VERSION_PATTERN =
            "\\d+\\.\\d+\\.\\d+(?:-(?:pre|rc)\\d+)?|b\\d+\\.\\d+\\.\\d+|\\d{2}w\\d{2}[a-z](?:_or_[a-z])?|[\\w.-]+";

    private static final Pattern SIMPLE_VERSION_PATTERN = Pattern.compile("^(\\d+(?:\\.\\d+){1,2})$");
    private static final Pattern EXACT_PATTERN = Pattern.compile("^(" + VERSION_PATTERN + ")$");
    private static final Pattern RANGE_PATTERN = Pattern.compile("^(" + VERSION_PATTERN + ")-(" + VERSION_PATTERN + ")$");
    private static final Pattern EQUAL_PATTERN = Pattern.compile("^=(" + VERSION_PATTERN + ")$");
    private static final Pattern TILDE_PATTERN = Pattern.compile("^~(" + VERSION_PATTERN + ")$");
    private static final Pattern CARET_PATTERN = Pattern.compile("^\\^(" + VERSION_PATTERN + ")$");
    private static final Pattern COMPARISON_PATTERN = Pattern.compile("^(>=|<=|>|<)\\s*(" + VERSION_PATTERN + ")$");
    private static final Pattern MAVEN_RANGE_PATTERN = Pattern.compile("^[\\[(]([\\w.,\\s-]+)[])]$");
    private static final Pattern COMPOSITE_PATTERN = Pattern.compile("^(.+?)\\s+(.+)$");

    public static one.pkg.modpublish.version.constraint.VersionConstraint parse(String constraintStr) throws IllegalArgumentException {
        String trimmed = constraintStr.trim();
        if (trimmed.isEmpty()) {
            throw new IllegalArgumentException("Version constraint cannot be empty");
        }

        String normalized = trimmed.replaceAll("([><=])\\s+", "$1");

        Matcher simpleMatcher = SIMPLE_VERSION_PATTERN.matcher(normalized);
        if (simpleMatcher.matches()) {
            return new ExactVersionConstraint(simpleMatcher.group(1));
        }

        Matcher equalMatcher = EQUAL_PATTERN.matcher(normalized);
        if (equalMatcher.matches()) return new ExactVersionConstraint(equalMatcher.group(1));

        Matcher rangeMatcher = RANGE_PATTERN.matcher(normalized);
        if (rangeMatcher.matches()) {
            Version min = new Version(rangeMatcher.group(1));
            Version max = new Version(rangeMatcher.group(2));
            return new RangeConstraint(min, max, true, true, trimmed);
        }

        Matcher tildeMatcher = TILDE_PATTERN.matcher(normalized);
        if (tildeMatcher.matches()) return new TildeConstraint(trimmed);

        Matcher caretMatcher = CARET_PATTERN.matcher(normalized);
        if (caretMatcher.matches()) return new CaretConstraint(trimmed);

        Matcher comparisonMatcher = COMPARISON_PATTERN.matcher(normalized);
        if (comparisonMatcher.matches()) {
            return parseComparison(comparisonMatcher.group(1), comparisonMatcher.group(2), trimmed);
        }

        Matcher mavenRangeMatcher = MAVEN_RANGE_PATTERN.matcher(normalized);
        if (mavenRangeMatcher.matches()) {
            return parseMavenRange(trimmed, mavenRangeMatcher.group(1));
        }

        Matcher compositeMatcher = COMPOSITE_PATTERN.matcher(normalized);
        if (compositeMatcher.matches()) {
            try {
                VersionConstraint first = parse(compositeMatcher.group(1));
                VersionConstraint second = parse(compositeMatcher.group(2));
                return new CompositeConstraint(List.of(first, second), trimmed);
            } catch (Exception e) {
                throw new IllegalArgumentException("Unable to parse version constraint: " + constraintStr);
            }
        }

        Matcher exactMatcher = EXACT_PATTERN.matcher(normalized);
        if (exactMatcher.matches()) return new ExactVersionConstraint(exactMatcher.group(1));

        throw new IllegalArgumentException("Unable to parse version constraint: " + constraintStr);
    }

    private static VersionConstraint parseComparison(String operator, String versionStr, String original)
            throws IllegalArgumentException {
        Version version = new Version(versionStr);
        return switch (operator) {
            case ">=" -> new RangeConstraint(version, null, true, false, original);
            case "<=" -> new RangeConstraint(null, version, false, true, original);
            case ">" -> new RangeConstraint(version, null, false, false, original);
            case "<" -> new RangeConstraint(null, version, false, false, original);
            default -> throw new IllegalArgumentException("Unknown comparison operator: " + operator);
        };
    }

    private static VersionConstraint parseMavenRange(String original, String content) {
        boolean includeMin = original.startsWith("[");
        boolean includeMax = original.endsWith("]");

        int end = content.length();
        while (end > 0 && content.charAt(end - 1) == ',') {
            end--;
        }
        String sliced = content.substring(0, end);

        int commaIndex = sliced.indexOf(',');

        if (commaIndex == -1) {
            if (original.endsWith(",)")) {
                Version min = new Version(sliced.trim());
                return new RangeConstraint(min, null, includeMin, false, original);
            } else {
                return new ExactVersionConstraint(sliced.trim());
            }
        }

        int nextCommaIndex = sliced.indexOf(',', commaIndex + 1);

        if (nextCommaIndex == -1) {
            String part1 = sliced.substring(0, commaIndex).trim();
            String part2 = sliced.substring(commaIndex + 1).trim();

            Version min = new Version(part1);
            Version max = part2.isEmpty() ? null : new Version(part2);
            return new RangeConstraint(min, max, includeMin, includeMax, original);
        } else {
            String[] parts = sliced.split(",");
            String part1 = parts[0].trim();
            String partLast = parts[parts.length - 1].trim();

            Version min = new Version(part1);
            Version max = partLast.isEmpty() ? null : new Version(partLast);
            RangeConstraint range = new RangeConstraint(min, max, includeMin, includeMax, original);

            List<VersionConstraint> constraints = new ArrayList<>();
            constraints.add(range);

            for (int i = 1; i < parts.length - 1; i++) {
                String mid = parts[i].trim();
                if (!mid.isEmpty()) {
                    constraints.add(new ExactVersionConstraint(mid));
                }
            }

            return new OrConstraint(constraints, original);
        }
    }
}

File: src/main/kotlin/one/pkg/modpublish/util/version/constraint/VersionConstraintParser.kt
/*
 * Copyright (C) 2025 404Setup (https://github.com/404Setup)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package one.pkg.modpublish.util.version.constraint

import one.pkg.modpublish.util.version.Version

object VersionConstraintParser {
    private const val VERSION_PATTERN =
        "\\d+\\.\\d+\\.\\d+(?:-(?:pre|rc)\\d+)?|b\\d+\\.\\d+\\.\\d+|\\d{2}w\\d{2}[a-z](?:_or_[a-z])?|[\\w.-]+"

    private val SIMPLE_VERSION_PATTERN = Regex("^(\\d+(?:\\.\\d+){1,2})$")
    private val EXACT_PATTERN = Regex("^($VERSION_PATTERN)$")
    private val RANGE_PATTERN = Regex("^($VERSION_PATTERN)-($VERSION_PATTERN)$")
    private val EQUAL_PATTERN = Regex("^=($VERSION_PATTERN)$")
    private val TILDE_PATTERN = Regex("^~($VERSION_PATTERN)$")
    private val CARET_PATTERN = Regex("^\\^($VERSION_PATTERN)$")
    private val COMPARISON_PATTERN = Regex("^(>=|<=|>|<)\\s*($VERSION_PATTERN)$")
    private val MAVEN_RANGE_PATTERN = Regex("^[\\[(]([\\w.,-]+)[])]$")
    private val COMPOSITE_PATTERN = Regex("^(.+?)\\s+(.+)$")

    @Throws(IllegalArgumentException::class)
    fun parse(constraintStr: String): VersionConstraint {
        val trimmed = constraintStr.trim()
        require(trimmed.isNotEmpty()) { "Version constraint cannot be empty" }

        // Check for simple version pattern first (e.g., 1.12.2)
        SIMPLE_VERSION_PATTERN.matchEntire(trimmed)?.let {
            return ExactVersionConstraint(it.groupValues[1])
        }

        EQUAL_PATTERN.matchEntire(trimmed)?.let { return ExactVersionConstraint(it.groupValues[1]) }
        RANGE_PATTERN.matchEntire(trimmed)?.let {
            val min = Version(it.groupValues[1])
            val max = Version(it.groupValues[2])
            return RangeConstraint(min, max, includeMin = true, includeMax = true, original = trimmed)
        }
        TILDE_PATTERN.matchEntire(trimmed)?.let { return TildeConstraint(trimmed) }
        CARET_PATTERN.matchEntire(trimmed)?.let { return CaretConstraint(trimmed) }
        COMPARISON_PATTERN.matchEntire(trimmed)?.let {
            return parseComparison(it.groupValues[1], it.groupValues[2], trimmed)
        }
        MAVEN_RANGE_PATTERN.matchEntire(trimmed)?.let {
            return parseMavenRange(trimmed, it.groupValues[1])
        }
        COMPOSITE_PATTERN.matchEntire(trimmed)?.let {
            return try {
                val first = parse(it.groupValues[1])
                val second = parse(it.groupValues[2])
                CompositeConstraint(listOf(first, second), trimmed)
            } catch (_: Exception) {
                throw IllegalArgumentException("Unable to parse version constraint: $constraintStr")
            }
        }

        // Fallback to generic EXACT_PATTERN for other cases
        EXACT_PATTERN.matchEntire(trimmed)?.let { return ExactVersionConstraint(it.groupValues[1]) }

        throw IllegalArgumentException("Unable to parse version constraint: $constraintStr")
    }

    @Throws(IllegalArgumentException::class)
    private fun parseComparison(operator: String, versionStr: String, original: String): VersionConstraint {
        val version = Version(versionStr)
        return when (operator) {
            ">=" -> RangeConstraint(version, null, includeMin = true, includeMax = false, original)
            "<=" -> RangeConstraint(null, version, includeMin = false, includeMax = true, original)
            ">" -> RangeConstraint(version, null, includeMin = false, includeMax = false, original)
            "<" -> RangeConstraint(null, version, includeMin = false, includeMax = false, original)
            else -> throw IllegalArgumentException("Unknown comparison operator: $operator")
        }
    }

    private fun parseMavenRange(original: String, content: String): VersionConstraint {
        val includeMin = original.startsWith("[")
        val includeMax = original.endsWith("]")

        val parts = content.split(",")
        return when (parts.size) {
            1 -> {
                if (original.endsWith(",)")) {
                    val min = Version(parts[0])
                    RangeConstraint(min, null, includeMin, includeMax = false, original)
                } else {
                    ExactVersionConstraint(parts[0])
                }
            }

            2 -> {
                val min = Version(parts[0])
                val max = if (parts[1].isEmpty()) null else Version(parts[1])
                RangeConstraint(min, max, includeMin, includeMax, original)
            }

            else -> {
                val constraints = parts.filter { it.isNotEmpty() }.map { ExactVersionConstraint(it) }
                OrConstraint(constraints, original)
            }
        }
    }
}
